### ISSUE TITLE
Property 'xxx' with value '/tmp/bug.txt' cannot be serialized with RegularFileProperty

### ISSUE THREAD

**Issue by szpak** [gradle/gradle#12351]
### Expected Behavior
I would like to be able to set an arbitrary file or directory location to `RegularFileProperty` defined as `@Input` (even if there is some **other** field in that task defined with `@OutputFile` or `@OutputDirectory` which should not be a problem).

### Current Behavior
```
Execution failed for task ':bug3'.
> Unable to store input properties for task ':bug3'. Property 'defaultFileForHistoryData'
with value '/tmp/RegularFilePropertyInputPlugin/build/bug.txt' cannot be serialized.
...
Caused by: java.lang.ClassCastException: class java.io.File cannot be cast to
class java.lang.String (java.io.File and java.lang.String are in module java.base of loader 'bootstrap')
```

### Context
I try to remove deprecation warnings in my [gradle-pitest-plugin](https://github.com/szpak/gradle-pitest-plugin/issues/155) with Gradle 6 about using `File` and `@Input` together. In one case, in my plugin configuration, there is a file pointed which location (not context - I cannot used `@InputFile` - file might not exist) will be used during the task execution.

### Steps to Reproduce 
1. Add the following to `build.gradle`:

```
class Bug3Task extends JavaExec {

    @OutputFile
    @Optional
    final RegularFileProperty reportFile   //required to trigger the issue, could be DirectoryProperty as well

    @Input
    final RegularFileProperty defaultFileForHistoryData

    Bug3Task() {
        defaultFileForHistoryData = project.objects.fileProperty()
        reportFile = project.objects.fileProperty()
    }

    @Override
    void exec() {
        logger.lifecycle("Value: ${defaultFileForHistoryData.get().asFile}" )
    }
}

class Bug3Plugin implements Plugin<Project> {

    @Override
    void apply(Project project) {
        project.getTasks().register("bug3", Bug3Task, { task ->
            task.defaultFileForHistoryData.set(new File(project.buildDir, "bug.txt"));
        });
    }
}

apply plugin: Bug3Plugin
```

2. Call `./gradlew bug3`

### Your Environment
Gradle 5.6.1 and 6.2.1. Fedora Linux. OpenJDK 11.0.6.

### Longer stactrace

```
$ gw bug3 -s
> Task :bug3 FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':bug3'.
> Unable to store input properties for task ':bug3'. Property 'defaultFileForHistoryData' with value '/tmp/RegularFilePropertyInputPlugin/build/bug.txt' cannot be serialized.

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':bug3'.
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:38)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.api.UncheckedIOException: Unable to store input properties for task ':bug3'. Property 'defaultFileForHistoryData' with value '/tmp/RegularFilePropertyInputPlugin/build/bug.txt' cannot be serialized.
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$fingerprintInputProperties$2(CaptureStateBeforeExecutionStep.java:178)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.visitInputProperties(ExecuteActionsTaskExecuter.java:307)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.fingerprintInputProperties(CaptureStateBeforeExecutionStep.java:169)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.captureExecutionState(CaptureStateBeforeExecutionStep.java:149)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$captureExecutionStateOp$1(CaptureStateBeforeExecutionStep.java:104)
        at org.gradle.internal.execution.steps.BuildOperationStep$1.call(BuildOperationStep.java:40)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.internal.execution.steps.BuildOperationStep.operation(BuildOperationStep.java:37)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.captureExecutionStateOp(CaptureStateBeforeExecutionStep.java:103)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$execute$0(CaptureStateBeforeExecutionStep.java:78)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:78)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:192)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        ... 23 more
Caused by: java.lang.ClassCastException: class java.io.File cannot be cast to class java.lang.String (java.io.File and java.lang.String are in module java.base of loader 'bootstrap')
        at org.gradle.internal.snapshot.impl.DefaultValueSnapshotter$ValueSnapshotVisitor.managedImmutableValue(DefaultValueSnapshotter.java:287)
        at org.gradle.internal.snapshot.impl.DefaultValueSnapshotter$ValueSnapshotVisitor.managedImmutableValue(DefaultValueSnapshotter.java:228)
        at org.gradle.internal.snapshot.impl.DefaultValueSnapshotter.processValue(DefaultValueSnapshotter.java:152)
        at org.gradle.internal.snapshot.impl.DefaultValueSnapshotter.snapshot(DefaultValueSnapshotter.java:54)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$fingerprintInputProperties$2(CaptureStateBeforeExecutionStep.java:173)
        ... 56 more
```

**Comment by wolfs** [comment]
This should work:
```
class Bug3Task extends JavaExec {

    @OutputFile
    @Optional
    final RegularFileProperty reportFile   //required to trigger the issue, could be DirectoryProperty as well

    @Internal
    final RegularFileProperty defaultFileForHistoryData

    @Input
    String getDefaultFileForHistoryDataPath() {
		defaultFileForHistoryData.asFile.get().absolutePath
	}

    Bug3Task() {
        defaultFileForHistoryData = project.objects.fileProperty()
        reportFile = project.objects.fileProperty()
    }

    @Override
    void exec() {
        logger.lifecycle("Value: ${defaultFileForHistoryData.get().asFile}" )
    }
}
```

A `RegularFileProperty` should not be used `@Input`, only at file input. If your task only cares about the path, then do the above. `RegularFileProperty.locationOnly` could be made to work better to produce a `Provider<String>` maybe.

**Comment by szpak** [comment]
Thanks Stefan for the workaround! It works.

I wonder, if it wasn't be worth to mention that somewhere in the documentation?

**Comment by adammurdoch** [comment]
We can certainly make the error message much clearer when this happens.

**Comment by vlsi** [comment]
TL;DR: If you see `Property 'propertyName' with value '/file/path' cannot be serialized` with using `RegularFileProperty`, then you need to replace `@Input` with `@InputFile`.

**Comment by netvl** [comment]
BTW, same is true for `DirectoryProperty` as well (cannot be used with `@Input`).

**Comment by vlsi** [comment]
I think this is still relevant

**Comment by wolfs** [comment]
Does this PR fix this: https://github.com/gradle/gradle/pull/21153 ?

**Comment by szpak** [comment]
@wolfs The error should be much better with #21153, so I think it does. @vlsi WDYT about covered cases?

**Comment by vlsi** [comment]
Indeed, #21153 resolved this
